### PR TITLE
[#957] Login on resuming form

### PR DESCRIPTION
**Pull request by SilviaAmAm**
Part of #957

### CODE REVIEW

File: src/openforms/authentication/contrib/demo/plugin.py [hunk 1]
```diff
@@ -56,13 +56,17 @@ def start_login(
...
         self, request: HttpRequest, form: Form
     ) -> Union[str, HttpResponse]:
-        submited = self.form_class(request.POST)
-        if not submited.is_valid():
+        submitted = self.form_class(request.POST)
+        if not submitted.is_valid():
             return HttpResponseBadRequest("invalid data")
 
-        request.session[self.provides_auth] = submited.cleaned_data[self.form_field]
+        request.session["form_auth"] = {
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/demo/plugin.py:63]: session key is defined as a constant in https://github.com/open-formulieren/open-forms/blob/master/src/openforms/authentication/constants.py#L5

File: src/openforms/authentication/contrib/demo/tests/test_plugin.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def test_login_bsn_and_baseclass(self):
         self.assertEqual(response.status_code, 405)
 
         # good
-        self.assertNotIn("bsn", self.client.session)
+        self.assertNotIn("form_auth", self.client.session)
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/demo/tests/test_plugin.py:58]: please use the constant

File: src/openforms/authentication/contrib/demo/tests/test_plugin.py [hunk 3]
```diff
@@ -64,8 +64,9 @@ def test_login_bsn_and_baseclass(self):
         self.assertEqual(response.status_code, 302)
         self.assertEqual(response["Location"], "http://foo.bar")
 
-        self.assertIn("bsn", self.client.session)
-        self.assertIn(self.client.session[AuthAttribute.bsn], "111222333")
+        self.assertIn("form_auth", self.client.session)
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/demo/tests/test_plugin.py:67]: please use the constant

File: src/openforms/authentication/contrib/demo/tests/test_plugin.py [hunk 4]
```diff
@@ -64,8 +64,9 @@ def test_login_bsn_and_baseclass(self):
...
+        self.assertEqual("bsn", self.client.session["form_auth"]["attribute"])
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/demo/tests/test_plugin.py:68]: please use `AuthAttribute.bsn` from the constants

File: src/openforms/authentication/contrib/eherkenning/views.py [hunk 5]
```diff
@@ -13,7 +13,8 @@
 from furl import furl
 from onelogin.saml2.errors import OneLogin_Saml2_ValidationError
 
-from openforms.authentication.contrib.digid.mixins import AssertionConsumerServiceMixin
+from ...constants import AuthAttribute
+from ...contrib.digid.mixins import AssertionConsumerServiceMixin
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/eherkenning/views.py:17]:
```suggestion
from ..digid.mixins import AssertionConsumerServiceMixin
```
?

File: src/openforms/submissions/tests/test_endpoints.py [hunk 6]
```diff
@@ -153,3 +154,185 @@ def test_token_valid_on_same_day(self):
...
+
+    def test_redirects_to_auth_if_form_requires_login(self):
+        form = FormFactory.create()
+        form_step = FormStepFactory.create(
+            form=form, form_definition__login_required=True
+        )
+        submission = SubmissionFactory.create(form=form, auth_plugin="digid")
+        SubmissionStepFactory.create(
+            submission=submission, form_step=form_step, data={"foo": "bar"}
+        )
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_endpoints.py:166]: think you can shortcut this a lot with either `SubmissionFactory.create(form__generate_minimal_setup=True, form__formstep__form_definition__login_required=True)` or `SubmissionFactory.from_components(...)`

File: src/openforms/submissions/tests/test_endpoints.py [hunk 7]
```diff
@@ -153,3 +154,185 @@ def test_token_valid_on_same_day(self):
...
+        expected_redirect_url = furl(f"http://testserver/auth/{form.slug}/digid/start")
+        expected_redirect_url.args["next"] = f"http://testserver{endpoint}"
+
+        response = self.client.get(endpoint)
+
+        self.assertRedirects(
+            response, expected_redirect_url.url, fetch_redirect_response=False
+        )
+        self.assertNotIn("form-submissions", self.client.session)
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_endpoints.py:183]: please use the constant defined for this session key

File: src/openforms/submissions/views.py [hunk 8]
```diff
@@ -32,13 +30,55 @@ def get_redirect_url(self, submission_uuid: uuid, token: str, *args, **kwargs):
...
+
+        # No login required, skip authentication
+        if not submission.form.login_required:
+            add_submmission_to_session(submission, self.request.session)
+            return form_resume_url.url
+
+        # Login IS required. Check if the user has already logged in.
+        # This is done by checking if the authentication details are in the session and
+        # if they match those in the saved submission.
+        if "form_auth" in self.request.session:
```
Comment (sergei-maertens) [inline src/openforms/submissions/views.py:50]: please use the constant

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
Can we also rename the `test_endpoints` module to something with a proper meaningful name?

File: src/openforms/authentication/contrib/eherkenning/views.py (hunk 5 above)
Comment (SilviaAmAm) [reply src/openforms/authentication/contrib/eherkenning/views.py:17]: No idea what happened here :sweat_smile:

File: src/openforms/submissions/tests/test_resume_form_view.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+            },
+        )
+        expected_redirect_url = furl(submission.form_url)
+        expected_redirect_url /= "stap"
+        expected_redirect_url /= form_step.form_definition.slug
+        expected_redirect_url.args["submission_uuid"] = submission.uuid
+
+        # Add form_auth to session, as the authentication plugin would do it
+        session = self.client.session
+        session["form_auth"] = {
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_resume_form_view.py]: Use constant

File: src/openforms/submissions/tests/test_resume_form_view.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+            "submissions:resume",
+            kwargs={
+                "token": submission_resume_token_generator.make_token(submission),
+                "submission_uuid": submission.uuid,
+            },
+        )
+
+        # Add form_auth to session, as the authentication plugin would do it
+        session = self.client.session
+        session["form_auth"] = {
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_resume_form_view.py]: Use constant

File: src/openforms/submissions/tests/test_resume_form_view.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_resume_form_view.py]: Use constant

File: src/openforms/submissions/tests/test_resume_form_view.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
```
Comment (sergei-maertens) [inline src/openforms/submissions/tests/test_resume_form_view.py]: Use constant

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**